### PR TITLE
Add ORC reader tutorial

### PR DESCRIPTION
**Pull request by oliverhu**
Tutorial for the ORC reader: https://github.com/tensorflow/io/issues/1372

### CODE REVIEW

**Comment by oliverhu** [comment]
@kvignesh1420

File: docs/tutorials/orc.ipynb [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+        "dataset = tf.data.Dataset.zip((feature_dataset, label_dataset))\n",
+        "dataset = dataset.batch(1)\n",
+        "\n",
+        "def pack_features_vector(features, labels):\n",
+        "    \"\"\"Pack the features into a single array.\"\"\"\n",
+        "    features = tf.stack(list(features), axis=1)\n",
+        "    return features, labels\n",
+        "\n",
+        "dataset = dataset.map(pack_features_vector)\n",
+        "\n",
```
Comment (kvignesh1420) [inline docs/tutorials/orc.ipynb:283]: can we separate these steps into a different section on dataset pre-processing?

**Comment by oliverhu** [comment]
@kvignesh1420  updated

File: docs/tutorials/orc.ipynb [hunk 2]
```diff
@@ -0,0 +1,398 @@
+{
```
Comment (kvignesh1420) [inline docs/tutorials/orc.ipynb:1]: <p>Please add a reference to the ORC documentation.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/io/pull/1465/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (kvignesh1420) [inline docs/tutorials/orc.ipynb:1]: <p>Packages should be lowercase (packages)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/io/pull/1465/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (kvignesh1420) [inline docs/tutorials/orc.ipynb:1]: <p>This section can be renamed to "dataset pre-processing".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/io/pull/1465/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (kvignesh1420) [inline docs/tutorials/orc.ipynb:1]: <p>I think this feature processing step can be modified to use the keras pre-processing layers.</p><p>For ex: - https://www.tensorflow.org/tutorials/structured_data/preprocessing_layers#categorical_columns</p><p>https://www.tensorflow.org/tutorials/structured_data/preprocessing_layers#numeric_columns</p><p> </p><p>Please modify the next cells as required.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/io/pull/1465/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (kvignesh1420) [inline docs/tutorials/orc.ipynb:1]: <p>This can be a new section titled "Build, compile and train the model"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/io/pull/1465/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

**Review by kvignesh1420 — CHANGES_REQUESTED** [review]
@oliverhu thanks for putting this together. Please check the comments for modifications.

File: docs/tutorials/orc.ipynb (hunk 2 above)
Comment (oliverhu) [reply docs/tutorials/orc.ipynb:1]: <p>This section is model building. i can add a subsection for data preprocessing.</p>

Comment (oliverhu) [reply docs/tutorials/orc.ipynb:1]: <p>keras preprocessing is experimental. It is better to use non experimental features in tutorial so it can be long lasting. We can update it later?</p>

Comment (kvignesh1420) [reply docs/tutorials/orc.ipynb:1]: <p>This section title can be removed. Please check the comments above and below.</p>

Comment (kvignesh1420) [reply docs/tutorials/orc.ipynb:1]: <p>okay sure.</p>

File: docs/tutorials/orc.ipynb [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
```
Comment (kvignesh1420) [inline docs/tutorials/orc.ipynb:1]: <p>I think it would be better if we directly print the dataset element:</p><pre class="ql-syntax" spellcheck="false">for item in dataset.take(1):
    print(item)

</pre><p>This reduces the lengthy syntax and the reader's can gain more information on the tensor.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/io/pull/1465/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (kvignesh1420) [inline docs/tutorials/orc.ipynb:1]: <p>duplicate</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/io/pull/1465/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (kvignesh1420) [inline docs/tutorials/orc.ipynb:1]: <p>Make this a section, similar to <code>Setup</code>  for ease in navigation.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/io/pull/1465/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

File: docs/tutorials/orc.ipynb (hunk 2 above)
Comment (kvignesh1420) [reply docs/tutorials/orc.ipynb:1]: just FYI, even though they are experimental, they are quite stable and are widely used.

**Comment by kvignesh1420** [comment]
@oliverhu I think the notebook should contain the following sections:
- Overview
- Setup
- Create tensorflow_io dataset
- Build, compile and train a model
with relevant sub-sections. This will help the users navigate the tutorial easily. Thanks again!

**Comment by oliverhu** [comment]
@kvignesh1420 updated

**Comment by oliverhu** [comment]
thanks for the review!

**Comment by kvignesh1420** [comment]
@oliverhu thanks for the changes.

cc: @yongtang to take a look for additional suggestions (if any).

**Comment by yongtang** [comment]
@oliverhu I run though the notebook and it works fine. Though can you remove the outputs from the notebook? Once it is published to tensorflow.org, then user can run on their own so no need to have outputs.

**Comment by yongtang** [comment]
@oliverhu It's also worth to add a little description on the dataset and the model. I can follow the notebook but user may want to know a little bit of the background when then run through steps.

**Comment by oliverhu** [comment]
@yongtang  sure, I thought we need to keep the output to give user some leads how the output is supposed to look like.  Will update the description as well.

**Comment by oliverhu** [comment]
@yongtang updated

**Comment by oliverhu** [comment]
@yongtang good to merge now? 😄

**Comment by yongtang** [comment]
Add @tensorflow/docs-team to take a look.

File: docs/tutorials/orc.ipynb [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+        "    [\n",
+        "        tf.keras.layers.Dense(\n",
+        "            10, activation=tf.nn.relu, input_shape=(4,)\n",
+        "        ),  # input shape required\n",
+        "        tf.keras.layers.Dense(10, activation=tf.nn.relu),\n",
+        "        tf.keras.layers.Dense(3),\n",
+        "    ]\n",
+        ")\n",
+        "\n",
+        "model.compile(optimizer=\"adam\", loss=\"binary_crossentropy\", metrics=[\"accuracy\"])\n",
```
Comment (MarkDaoust) [inline docs/tutorials/orc.ipynb:316]: > loss=\"binary_crossentropy\"

BInary crossentrpy is not the right loss here. You probably want `CategoricalCrossentropy(from_logits=True)`.

File: docs/tutorials/orc.ipynb [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "TQvuE7OgVs1q"
+      },
+      "outputs": [],
+      "source": [
+        "@tf.function\n",
+        "def species_float_conversion(x):\n",
```
Comment (MarkDaoust) [inline docs/tutorials/orc.ipynb:250]: This function is really strange.

1. Returns floats -> should probably return ints
2. Result is one-based
3. It's not scalable:
    * It does a linear scan to find the value instead of a dictionary lookup.
    * It can't handle batches

Either use `tf.lookup` for a tensorflow equivalent of `dict` or use `tf.keras.layers.StringLookup`/`tf.keras.layers.experimental.preprocessing`:

```
lookup = tf.keras.layers.StringLookup(
    vocabulary=["virginica","versicolor","setosa"],
     num_oov_indices=0)

lookup(["virginica","versicolor","setosa"])
```
```
<tf.Tensor: shape=(3,), dtype=int64, numpy=array([0, 1, 2])>
```

```
lookup = tf.keras.layers.StringLookup(
    vocabulary=["virginica","versicolor","setosa"],
     num_oov_indices=0, invert=True)

lookup([0,1,2,3])
```
```
<tf.Tensor: shape=(4,), dtype=string, numpy=array([b'virginica', b'versicolor', b'setosa', b'[UNK]'], dtype=object)>
```

File: docs/tutorials/orc.ipynb [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+      "metadata": {
+        "id": "tToy0FoOWG-9"
+      },
+      "outputs": [],
+      "source": [
+        "model = tf.keras.Sequential(\n",
+        "    [\n",
+        "        tf.keras.layers.Dense(\n",
+        "            10, activation=tf.nn.relu, input_shape=(4,)\n",
+        "        ),  # input shape required\n",
```
Comment (MarkDaoust) [inline docs/tutorials/orc.ipynb:310]: > `# input shape required`

I'm pretty sure it's only recommended, not required.

**Review by MarkDaoust — CHANGES_REQUESTED** [review]
Mostly LGTM to me. Just a few comments to fix.

File: docs/tutorials/orc.ipynb (hunk 5 above)
Comment (oliverhu) [reply docs/tutorials/orc.ipynb:250]: Thanks for shedding lights on this!

**Comment by oliverhu** [comment]
@MarkDaoust  thanks for the review, good catches! updated, pls take another look

File: docs/tutorials/orc.ipynb [hunk 7]
```diff
@@ -0,0 +1,333 @@
+{
```
Comment (kvignesh1420) [inline docs/tutorials/orc.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.        init,
</pre><p>this should be <code>vocab_init</code>  ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/io/pull/1465/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (oliverhu) [reply docs/tutorials/orc.ipynb:1]: right

**Review by MarkDaoust — APPROVED** [review]
LGTM

**Review by kvignesh1420 — APPROVED** [review]
Thanks @oliverhu
